url542ae48576_tonyheadford_m68k_tgz-pJ8-m68k_MonitorJ8/src/m68k/cpu/CpuCore.java

```diff
@@ -0,0 +1,2574 @@
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+//  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+*/
+public abstract class CpuCore implements Cpu
+{
+	protected AddressSpace memory;
+	protected int[] data_regs;
+	protected int[] addr_regs;
+	protected int reg_pc;
+	protected int reg_sr;
+	protected int reg_usp;
+	protected int reg_ssp;
+	protected Operand[] srcHandlers;
+	protected Operand[] dstHandlers;
+	protected Operand srcEAHandler;
+	protected Operand dstEAHandler;
+	protected int currentInstructionAddress;
+	protected StringBuilder disasmBuffer;
+
+	public CpuCore()
+	{
+		data_regs = new int[8];
+		addr_regs = new int[8];
+
+		srcEAHandler = null;
+		dstEAHandler = null;
+
+		memory = null;
+
+		disasmBuffer = new StringBuilder(64);
+		initEAHandlers();
+	}
+
+	public void setAddressSpace(AddressSpace as)
+	{
+		this.memory = as;
+	}
+
+	public void reset()
+	{
+		//TODO: this has to be sent to all external devices - called by RESET instruction
+		//and during initialization (is this correct ?)
+
+		reg_ssp = memory.readLong(0);
+		addr_regs[7] = reg_ssp;
+		reg_pc = memory.readLong(4);
+		//supervisor mode, interrupts enabled
+		reg_sr = 0x2700;
+	}
+
+	public void stop()
+	{
+		//TODO: called by STOP instruction - should halt cpu and wait for interrupt
+	}
+
+	public int getDataRegisterByte(int reg)
+	{
+		return data_regs[reg] & 0x00ff;
+	}
+	public int getDataRegisterByteSigned(int reg)
+	{
+		return signExtendByte(data_regs[reg]);
+	}
+	public int getDataRegisterWord(int reg)
+	{
+		return data_regs[reg] & 0x0000ffff;
+	}
+	public int getDataRegisterWordSigned(int reg)
+	{
+		return signExtendWord(data_regs[reg]);
+	}
+	public int getDataRegisterLong(int reg)
+	{
+		return data_regs[reg];
+	}
+	public void setDataRegisterByte(int reg, int value)
+	{
+		data_regs[reg] = (data_regs[reg] & 0xffffff00) | (value & 0x00ff);
+	}
+	public void setDataRegisterWord(int reg, int value)
+	{
+		data_regs[reg] = (data_regs[reg] & 0xffff0000) | (value & 0x0000ffff);
+	}
+	public void setDataRegisterLong(int reg, int value)
+	{
+		data_regs[reg] = value;
+	}
+	// address registers
+	public int getAddrRegisterByte(int reg)
+	{
+		return addr_regs[reg] & 0x00ff;
+	}
+	public int getAddrRegisterByteSigned(int reg)
+	{
+		return signExtendByte(addr_regs[reg]);
+	}
+	public int getAddrRegisterWord(int reg)
+	{
+		return addr_regs[reg] & 0x0000ffff;
+	}
+	public int getAddrRegisterWordSigned(int reg)
+	{
+		return signExtendWord(addr_regs[reg]);
+	}
+	public int getAddrRegisterLong(int reg)
+	{
+		return addr_regs[reg];
+	}
+	public void setAddrRegisterByte(int reg, int value)
+	{
+		addr_regs[reg] = (addr_regs[reg] & 0xffffff00) | (value & 0x00ff);
+		if(reg == 7)
+		{
+			if(isSupervisorMode())
+			{
+				reg_ssp = addr_regs[reg];
+			}
+			else
+			{
+				reg_usp = addr_regs[reg];
+			}
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 136:147, 151:162.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.